internal/terraform/eval_variable.go

```diff
@@ -245,26 +302,49 @@ func evalVariableValidations(addr addrs.AbsInputVariableInstance, config *config
 		// we discard the marks now.
 		result, _ = result.Unmark()
 
-		if result.False() {
-			if expr != nil {
+		if result.True() {
+			continue
+		}
+
+		var errorMessage string
+		if !errorDiags.HasErrors() && errorValue.IsKnown() && !errorValue.IsNull() {
+			var err error
+			errorValue, err = convert.Convert(errorValue, cty.String)
+			if err != nil {
 				diags = diags.Append(&hcl.Diagnostic{
-					Severity: hcl.DiagError,
-					Summary:  errInvalidValue,
-					Detail:   fmt.Sprintf("%s\n\nThis was checked by the validation rule at %s.", validation.ErrorMessage, validation.DeclRange.String()),
-					Subject:  expr.Range().Ptr(),
+					Severity:    hcl.DiagError,
+					Summary:     "Invalid error message",
+					Detail:      fmt.Sprintf("Error message must be a string: %s.", tfdiags.FormatError(err)),
+					Subject:     validation.ErrorMessage.Range().Ptr(),
+					Expression:  validation.ErrorMessage,
+					EvalContext: hclCtx,
 				})
 			} else {
-				// Since we don't have a source expression for a root module
-				// variable, we'll just report the error from the perspective
-				// of the variable declaration itself.
-				diags = diags.Append(&hcl.Diagnostic{
-					Severity: hcl.DiagError,
-					Summary:  errInvalidValue,
-					Detail:   fmt.Sprintf("%s\n\nThis was checked by the validation rule at %s.", validation.ErrorMessage, validation.DeclRange.String()),
-					Subject:  config.DeclRange.Ptr(),
-				})
+				errorMessage = errorValue.AsString()
```

Comment:
Do you think we might want to run this through `strings.TrimSpace` to make it easier to write a string template that might otherwise generate extra leading and trailing whitespace? I'm not thinking of any situations where extra space there would be useful, and in particular with a heredoc string the result will _always_ end with a newline, which I think would make the format string below end up including three `\n` in sequence: one from the heredoc, and then the two hard-coded in the format string.

Reply:
I had noticed the extra heredoc newline in manual testing but didn't think about trimming space to address it 🤦 Thanks!